### 关于log.panic()

有的时候程序一直崩溃，你想获得尽可能多的错误信息。

在这种困难的情况下，你可以考虑使用`log.Panic()`，这就是本节中的程序`logPanic.go`要展示的。

这个程序是这样的：

```go
package main
import (
	"fmt"
	"log"
	"log/syslog"
)
func main() {
	sysLog, err := syslog.New(syslog.LOG_ALERT|syslog.LOG_MAIL, "Some program!")
	if err != nil {
		log.Fatal(err) 
	} else {
		log.SetOutput(sysLog) 
	}
	log.Panic(sysLog)
	fmt.Println("Will you see this?") 
}
```

在macOS Mojave上执行这个程序，会生成以下的输出：

```shell
$ go run logPanic.go
panic: &{17 Some program! iMac.local {0 0} 0xc0000b21e0} 
goroutine 1 [running]:
log.Panic(0xc00004ef68, 0x1, 0x1)
	/usr/local/Cellar/go/1.11.4/libexec/src/log/log.go:326 +0xc0 
main.main()
	/Users/mtsouk/Desktop/mGo2nd/Mastering-Go-Second- Edition/ch01/logPanic.go:17 +0xd6
exit status 2
```

在Debian系统下，Go版本1.3.3，执行同样的代码会生成以下输出：

```shell
$ go run logPanic.go
panic: &{17 Some program! mail {0 0} 0xc2080400e0} 
goroutine 16 [running]:
runtime.panic(0x4ec360, 0xc208000320)
	/usr/lib/go/src/pkg/runtime/panic.c:279 +0xf5 
log.Panic(0xc208055f20, 0x1, 0x1)
	/usr/lib/go/src/pkg/log/log.go:307 +0xb6 
main.main()
	/home/mtsouk/Desktop/masterGo/ch/ch1/code/logPanic.go:17 +0x169 
goroutine 17 [runnable]:
runtime.MHeap_Scavenger()
	/usr/lib/go/src/pkg/runtime/mheap.c:507 
runtime.goexit()
	/usr/lib/go/src/pkg/runtime/proc.c:1445 
goroutine 18 [runnable]:
bgsweep()
	/usr/lib/go/src/pkg/runtime/mgc0.c:1976 
runtime.goexit()
	/usr/lib/go/src/pkg/runtime/proc.c:1445 
goroutine 19 [runnable]:
runfinq()
	/usr/lib/go/src/pkg/runtime/mgc0.c:2606 
runtime.goexit()
	/usr/lib/go/src/pkg/runtime/proc.c:1445 
exit status 2
```

因此，函数`log.Panic()`会生成底层的错误信息，这些信息会帮助你解决程序中的难题。

和函数`log.Fatal()`一样，`log.Panic()`也会发送一条日志到对应的日志文件，并且终止程序的执行。

### 